addons/mail/static/src/discuss/google_translate/common/message_patch.js

```diff
@@ -0,0 +1,27 @@
+/* @odoo-module */
+
+import { Message } from "@mail/core/common/message";
+import { patch } from "@web/core/utils/patch";
+import { markup } from "@odoo/owl";
+
+patch(Message.prototype, {
+    translatable() {
+        return (
+            this.store.hasMessageTranslationFeature &&
+            !this.env.inChatWindow &&
+            !this.message.isSelfAuthored
+        );
+    },
+    async onClickToggleTranslation() {
+        if (!this.message.translationValue) {
+            const { value, source, error } = await this.rpc("/discuss/message/translate", {
+                message_id: this.message.id,
+            });
+            this.message.translationValue = value && markup(value);
+            this.message.translationSource = source && markup(source);
+            this.message.translationErrors = error && markup(error);
```

Comment:
You should document all these props in `Message` model in `message_model.js`.

Reply:
@alexkuhn 
Updated the following file: addons/mail/static/src/discuss/google_translate/common/message_patch.js
Is it what you expected or did mean updating the original message_model.js file ?
Also I decided to initialize thes attributes to null instead of undefined but the latter might be the right choice for the sake of consistency. What do you think ?

Reply:
It's preferred to keep a single class per file, so that we don't have to deal with inconveniences like: hard to find filepath where class is defined/patched, it's hard to tell which class owns the method at a glance, etc.

I'm okay with splitting concerns of components and services per feature and per environment, but not for models. For me it is much better to define them in original model file, so `core/common/message_model.js`. @seb-odoo would disagree with me: https://github.com/odoo/odoo/pull/130709#discussion_r1315001000

I think `patch()` is a significant maintenance cost that I don't want to pay when there's no functional motivation for it. Encapsulating the data in a feature sounds good sometimes, but I really don't want to incite making tons of features and then we loose understanding of the whole code due to huge complexity of features working together in a way that we can't grasp just by looking at the code at a glance... Yes, a big amount of features is also a significant maintenance cost, micro-services was a trend a few years ago and now people did experience huge drawbacks of such an architecture: https://www.shopify.com/enterprise/disadvantages-microservices

To sump up, it's not easy to find a right balance of amount of features, and how to split it in features. There are pros and cons to _too few_ or _too many_ feature encapsulation in the system. It's important to balance the amount of features to our needs, and usually this means that defining a feature has a cost that has a clear gain right now. If that's not clear, then it's much much better to not pay the price.

=====================

For initializing the attributes, it's shorter with just default JS value which is `undefined`:
```js
class Message extends Record {
   translationValue;
}
```

Also `undefined` has more desirable properties than `null`. For example, default params are set when `undefined` is provided, whereas `null` don't. This means `null` is a sentinel value whereas `undefined` really means "unset".

Reply:
Complete and detailed explanation as usual, thank you for your input Alex!